File: apps/web/src/types/orchestration.ts
/**
 * TypeScript types for the Test Shard Orchestration API.
 *
 * Mirrors the canonical OpenAPI definitions in apps/server/api/openapi.yaml.
 *
 * The composite identity tuple
 *   (repository, commit_sha, gh_run_id, name, gh_run_attempt)
 * addresses every orchestration resource on the API. The server's internal
 * UUIDs (run_id, lease_id) are not part of the worker-facing contract and
 * remain optional in these shapes.
 */

// ─── Identity ──────────────────────────────────────────────────────────────

export interface CompositeIdentity {
  repository: string;
  commit_sha: string;
  gh_run_id: string;
  name: string;
  /** Defaults to "1" on the server when omitted. */
  gh_run_attempt?: string;
  /** Test framework label. Server accepts "playwright" or "cypress". */
  framework?: 'playwright' | 'cypress';
  branch?: string;
  gh_pr_number?: number;
}

export interface WorkerIdentity {
  gh_job_name: string;
  gh_job_id: string;
}

// ─── Status & state enums ──────────────────────────────────────────────────

export type RunStatus = 'in_progress' | 'completed' | 'timed_out';

export type UnitState =
  | 'pending'
  | 'leased'
  | 'completed_pass'
  | 'completed_fail'
  | 'completed_skipped'
  | 'abandoned';

export type TestCaseStatus = 'passed' | 'failed' | 'skipped' | 'flaky' | 'timedOut' | 'interrupted';

// ─── Run snapshot (GET /orchestration/status response) ─────────────────────

export interface RunCounts {
  pending: number;
  leased: number;
  completed_pass: number;
  completed_fail: number;
  completed_skipped: number;
  abandoned: number;
  /**
   * Subset of completed_fail still within retest budget. The run is
   * `completed` only when this is zero alongside pending and leased.
   */
  retest_eligible?: number;
  /** Convenience field; not in the OpenAPI schema but commonly derived. */
  total?: number;
}

export interface RunSnapshot extends CompositeIdentity {
  status: RunStatus;
  total_units: number;
  started_at: string;
  /**
   * Bumped to now() on every successful checkout / complete. Combined
   * with `idle_timeout_ms` by the server-side reaper to decide when an
   * in-progress run transitions to `timed_out`.
   */
  last_activity_at: string;
  /** Inactivity window in milliseconds; run is reaped when no activity within this window. */
  idle_timeout_ms: number;
  terminal_at?: string | null;
  counts: RunCounts;
  /**
   * Per-dispatch-unit state, ordered by dispatch_seq. Each unit addresses
   * exactly one spec file. Returned by /begin (all pending) and /status.
   * Optional only for forward-compat with older server versions that did
   * not yet emit it.
   */
  units?: SnapshotUnit[];
}

/**
 * Per-spec attempt in the run-status payload, denormalized with the worker
 * name from its parent lease so the dashboard can render "spec X was run
 * by worker Y" without a per-row lookup. `test_cases` carries the
 * framework-specific per-test-case detail the worker supplied on /complete,
 * including attachments (screenshots, traces, etc.) that the dashboard
 * resolves through /files/{key}.
 */
export interface SnapshotAttempt {
  id: string;
  lease_id: string;
  spec_path: string;
  status: TestCaseStatus | null;
  actual_duration_ms: number | null;
  error_message: string | null;
  reported_at: string | null;
  late_report: boolean;
  expired: boolean;
  created_at: string;
  gh_job_name: string;
  gh_job_id: string;
  test_cases: SnapshotTestCase[] | null;
}

/**
 * Subset of SpecResultTestCase that the dashboard renders. Not every field
 * the worker can send is needed here — only the bits the per-attempt UI
 * surfaces. Extra unmapped fields are tolerated by TypeScript at runtime.
 */
export interface SnapshotTestCase {
  title: string;
  full_title: string;
  status: TestCaseStatus;
  retry_count?: number;
  duration_ms?: number | null;
  ordinal?: number;
  error_message?: string | null;
  error_stack?: string | null;
  attachments?: SnapshotTestCaseAttachments | null;
}

/**
 * The worker's `/complete` request rewrites local image paths into
 * { screenshots: [{ key, relative_path }] } before sending. The dashboard
 * resolves each `key` via /files/{key} (the public download path used by
 * Report Group views).
 */
export interface SnapshotTestCaseAttachments {
  screenshots?: Array<{ key: string; relative_path?: string | null }>;
}

/** Minimal lease info shown while a unit is in the leased state. */
export interface SnapshotCurrentLease {
  id: string;
  gh_job_name: string;
  gh_job_id: string;
  issued_at: string;
  deadline: string;
}

/**
 * One dispatch_units row in the run-status payload. Each unit addresses a
 * single spec file. `current_lease` is populated only while
 * `state === 'leased'`. `attempts` is oldest-first.
 */
export interface SnapshotUnit {
  id: string;
  dispatch_seq: number;
  spec_path: string;
  state: UnitState;
  lease_count: number;
  fail_count: number;
  outcome_set_at: string | null;
  current_lease: SnapshotCurrentLease | null;
  attempts: SnapshotAttempt[];
}

// ─── Per-unit / per-lease / per-attempt detail ─────────────────────────────

export interface OrchestrationUnit {
  unit_id: string;
  dispatch_seq: number;
  spec_path: string;
  state: UnitState;
  lease_count: number;
  fail_count: number;
}

export interface OrchestrationLease {
  /** Internal UUID; surfaced only on the UI's run-detail view, not over the worker API. */
  lease_id?: string;
  gh_job_name: string;
  gh_job_id: string;
  issued_at: string;
  deadline: string;
  released_at?: string | null;
  release_reason?: 'completed' | 'expired' | 'run_timed_out' | null;
}

export interface TestCaseDetail {
  title: string;
  full_title: string;
  status: TestCaseStatus;
  retry_count: number;
  duration_ms?: number | null;
  error_message?: string | null;
  error_stack?: string | null;
  annotations?: unknown[];
  attachments?: Record<string, unknown> | null;
  ordinal: number;
}

export interface OrchestrationAttempt {
  /** Internal UUID; FK back to OrchestrationLease. */
  lease_id: string;
  gh_job_name: string;
  gh_job_id: string;
  /**
   * The spec the attempt covers. Each row in the `attempts` table is keyed
   * by `(lease_id, spec_path)`; the worker contract sends one attempt row
   * per spec_path in its lease.
   */
  spec_path?: string;
  deadline: string;
  expired: boolean;
  late_report: boolean;
  /**
   * True when this attempt was issued as a retest dispatch (i.e. its parent
   * unit had already failed at least once). Mirrors the `is_retest` flag the
   * worker receives on /checkout and the corresponding event payload.
   */
  is_retest?: boolean;
  /**
   * Total accumulated failures on this attempt's dispatch unit at the time
   * the attempt was created. Surfaced primarily on retest attempts where
   * the orchestrator returns the prior fail_count alongside the dispatch.
   */
  fail_count?: number;
  status: TestCaseStatus | null;
  actual_duration_ms?: number | null;
  error_message?: string | null;
  error_stack?: string | null;
  test_cases?: TestCaseDetail[];
  reported_at?: string | null;
}

// ─── Cross-source divergence ───────────────────────────────────────────────

/**
 * A per-spec disagreement between the orchestration-recorded outcome and
 * the canonical artifact-derived outcome (consolidated test_cases). The UI
 * surfaces these so reviewers can immediately spot specs where the two
 * sources of truth differ.
 */
export interface Divergence {
  spec_path: string;
  orchestration_status: TestCaseStatus;
  artifact_status: TestCaseStatus;
}

// ─── Live progress events ──────────────────────────────────────────────────

export type OrchestrationEventType =
  | 'orchestration.run.started'
  | 'orchestration.unit.leased'
  | 'orchestration.unit.completed'
  | 'orchestration.lease.expired'
  | 'orchestration.run.completed'
  | 'orchestration.run.timed_out';

export interface RunStartedPayload {
  total_units: number;
  deadline: string;
  lease_timeout_ms: number;
}

export interface UnitLeasedPayload {
  gh_job_name: string;
  gh_job_id: string;
  unit_ids: string[];
  deadline: string;
  is_retest?: boolean;
}

export interface UnitCompletedPayload {
  unit_id: string;
  outcome: UnitState;
  late_report: boolean;
  attempts_count: number;
}

export interface LeaseExpiredPayload {
  gh_job_name: string;
  gh_job_id: string;
  released_at: string;
  reclaimed_unit_ids: string[];
}

export interface RunCompletedPayload {
  terminal_at: string;
  counts: RunCounts;
}

export interface RunTimedOutPayload {
  terminal_at: string;
  counts: RunCounts;
  abandoned_count: number;
}

export interface OrchestrationEventEnvelope<T extends OrchestrationEventType, P> {
  type: T;
  identity: CompositeIdentity;
  timestamp: string;
  payload: P;
}

export type OrchestrationRunStartedEvent = OrchestrationEventEnvelope<
  'orchestration.run.started',
  RunStartedPayload
>;
export type OrchestrationUnitLeasedEvent = OrchestrationEventEnvelope<
  'orchestration.unit.leased',
  UnitLeasedPayload
>;
export type OrchestrationUnitCompletedEvent = OrchestrationEventEnvelope<
  'orchestration.unit.completed',
  UnitCompletedPayload
>;
export type OrchestrationLeaseExpiredEvent = OrchestrationEventEnvelope<
  'orchestration.lease.expired',
  LeaseExpiredPayload
>;
export type OrchestrationRunCompletedEvent = OrchestrationEventEnvelope<
  'orchestration.run.completed',
  RunCompletedPayload
>;
export type OrchestrationRunTimedOutEvent = OrchestrationEventEnvelope<
  'orchestration.run.timed_out',
  RunTimedOutPayload
>;

/** Discriminated union of all orchestration event shapes delivered over the WebSocket. */
export type OrchestrationEvent =
  | OrchestrationRunStartedEvent
  | OrchestrationUnitLeasedEvent
  | OrchestrationUnitCompletedEvent
  | OrchestrationLeaseExpiredEvent
  | OrchestrationRunCompletedEvent
  | OrchestrationRunTimedOutEvent;

// ─── Helpers ───────────────────────────────────────────────────────────────

/**
 * Return a deterministic string key for a CompositeIdentity, suitable for
 * React Query cache keys, Map lookups, and routing. The shape mirrors the
 * five identity components in canonical order; missing optional values fall
 * back to the same defaults the server uses.
 */
export function compositeIdentityKey(id: CompositeIdentity): string {
  const attempt = id.gh_run_attempt ?? '1';
  return `${id.repository}/${id.commit_sha}/${id.gh_run_id}/${id.name}/${attempt}`;
}

/**
 * Return true when two identities address the same orchestration run. Uses
 * the canonical key form, so optional `gh_run_attempt` defaulting is consistent.
 */
export function compositeIdentityEquals(a: CompositeIdentity, b: CompositeIdentity): boolean {
  return compositeIdentityKey(a) === compositeIdentityKey(b);
}

/**
 * Worker /checkout response. Surfaces `is_retest` so a worker (and the UI
 * mocking a worker, in tests) can distinguish a normal first-pass dispatch
 * from a retest re-dispatch.
 */
export interface CheckoutResponse {
  units: Array<{
    unit_id: string;
    dispatch_seq: number;
    spec_path: string;
    fail_count?: number;
  }>;
  queue_empty: boolean;
  is_retest: boolean;
  lease_id?: string;
  deadline?: string;
}

/** Begin-run request body. Exposed for the dev/admin mutation hook. */
export interface BeginRunRequest extends CompositeIdentity {
  playwright_project?: string;
  lease_timeout_ms?: number;
  idle_timeout_ms?: number;
  retest_on_fail?: boolean;
  retest_budget?: number;
  dispatch_units: Array<{ spec_path: string }>;
}

File: apps/server/internal/orchestration/orchestration.go
// Package orchestration is the domain layer for the Test Shard Orchestration
// feature: stateless API addressing by composite identity, atomic dispatch of
// spec-file units to GitHub Actions worker shards, lease/run-level timeouts,
// per-unit retest, and live progress events.
//
// The package contains only types, enums, and small pure helpers. Database
// I/O lives in store.go; HTTP handlers, route wiring, and event publishing
// live in sibling packages.
package orchestration

import (
	"crypto/sha256"
	"encoding/json"
	"errors"
	"fmt"
	"strings"
	"time"

	"github.com/google/uuid"
)

// Framework label constants. Membership of a value in supportedFrameworks
// below implies the rest of the orchestration stack (queue, leases,
// retest, screenshot upload) is exercised against it; the
// orchestration_runs.framework CHECK constraint in the database mirrors
// this set and MUST be kept in sync when a new framework is added.
const (
	FrameworkPlaywright = "playwright"
	FrameworkCypress    = "cypress"
	FrameworkDetox      = "detox"
	FrameworkMaestro    = "maestro"

	// DefaultFramework is the value used when a request omits framework.
	// Kept at playwright for backward compatibility with consumers that
	// pre-date Cypress support.
	DefaultFramework = FrameworkPlaywright
)

// supportedFrameworks is the closed set of accepted framework labels.
// Lookup is membership-only; callers do not need to know the order.
var supportedFrameworks = map[string]struct{}{
	FrameworkPlaywright: {},
	FrameworkCypress:    {},
	FrameworkDetox:      {},
	FrameworkMaestro:    {},
}

// IsSupportedFramework reports whether s is one of the orchestration-
// supported framework labels. Empty input returns false — callers apply
// DefaultFramework before consulting this.
func IsSupportedFramework(s string) bool {
	_, ok := supportedFrameworks[s]
	return ok
}

// SupportedFrameworksList returns accepted framework labels in a stable
// order for error messages and docs.
func SupportedFrameworksList() []string {
	order := []string{FrameworkPlaywright, FrameworkCypress, FrameworkDetox, FrameworkMaestro}
	out := make([]string, 0, len(order))
	for _, f := range order {
		if _, ok := supportedFrameworks[f]; ok {
			out = append(out, f)
		}
	}
	return out
}

// Run status enum values.
const (
	RunStatusInProgress = "in_progress"
	RunStatusCompleted  = "completed"
	RunStatusTimedOut   = "timed_out"
)

// DispatchUnit state enum values.
const (
	UnitStatePending          = "pending"
	UnitStateLeased           = "leased"
	UnitStateCompletedPass    = "completed_pass"
	UnitStateCompletedFail    = "completed_fail"
	UnitStateCompletedSkipped = "completed_skipped"
	UnitStateAbandoned        = "abandoned"
)

// Lease release reason enum values.
const (
	LeaseReleaseCompleted   = "completed"
	LeaseReleaseExpired     = "expired"
	LeaseReleaseRunTimedOut = "run_timed_out"
)

// Attempt status enum values. Mirrors test_cases.status.
const (
	AttemptStatusPassed      = "passed"
	AttemptStatusFailed      = "failed"
	AttemptStatusSkipped     = "skipped"
	AttemptStatusFlaky       = "flaky"
	AttemptStatusTimedOut    = "timedOut"
	AttemptStatusInterrupted = "interrupted"
)

// Sentinel errors returned from the domain and store layers. Handlers map
// these to HTTP responses; the domain package is HTTP-agnostic.
var (
	// ErrConflict signals that a request conflicts with persisted state
	// (e.g. begin run with a different dispatch list, or checkout while the
	// worker already has an active lease).
	ErrConflict = errors.New("orchestration: conflict")

	// ErrNotFound signals that the addressed run / lease / unit does not
	// exist.
	ErrNotFound = errors.New("orchestration: not found")

	// ErrWorkerHasActiveLease signals a checkout from a worker that already
	// has an unreleased lease on the same run.
	ErrWorkerHasActiveLease = errors.New("orchestration: worker already has an active lease")

	// ErrRunNotInProgress signals that an operation requiring an active run
	// was issued against a run that has reached a terminal state.
	ErrRunNotInProgress = errors.New("orchestration: run is not in progress")

	// ErrPartialReport signals that a complete call covered fewer specs
	// than the lease's unit set.
	ErrPartialReport = errors.New("orchestration: partial report not allowed")

	// ErrUnknownLease signals that no lease exists for the given worker
	// identity on the given run (active or released).
	ErrUnknownLease = errors.New("orchestration: unknown lease for worker")
)

// CompositeIdentity is the public, stateless addressing tuple for an
// orchestration run. Mirrors the columns used by the existing report_groups
// table so the UI can join the two systems at query time without a foreign
// key.
type CompositeIdentity struct {
	Repository   string
	CommitSHA    string
	GHRunID      string
	Name         string
	GHRunAttempt string // defaults to "1" when omitted by the caller
	Branch       string
	GHPRNumber   *int
	Framework    string // pinned to the package-level Framework constant
}

// WorkerIdentity is the structured worker identifier carried on checkout and
// complete. Both fields are required and mirror reports.gh_job_name /
// reports.gh_job_id used by the per-shard upload flow.
type WorkerIdentity struct {
	GHJobName string
	GHJobID   string
}

// RunCounts is the materialized counter snapshot stored on orchestration_runs.
// The six state counters always sum to Total; RetestEligible is a derived
// subset of CompletedFail (units waiting to be re-leased) and is NOT included
// in the sum.
type RunCounts struct {
	Pending          int
	Leased           int
	CompletedPass    int
	CompletedFail    int
	CompletedSkipped int
	Abandoned        int
	RetestEligible   int
	Total            int
}

// Run is the in-memory view of an orchestration_runs row plus its
// materialized counters and owner.
type Run struct {
	ID                uuid.UUID // internal PK; never exposed in the API
	Identity          CompositeIdentity
	PlaywrightProject string
	LeaseTimeoutMs    int64
	// IdleTimeoutMs is the inactivity window: a run is reaped when no
	// checkout/complete has touched it for this many milliseconds.
	IdleTimeoutMs     int64
	RetestOnFail      bool
	RetestBudget      int
	Status            string
	Counts            RunCounts
	DispatchUnitsHash []byte
	StartedAt         time.Time
	// LastActivityAt is bumped to now() on every successful checkout and
	// complete. Combined with IdleTimeoutMs by the reaper to decide when
	// the run transitions to 'timed_out'.
	LastActivityAt   time.Time
	TerminalAt       *time.Time
	OwnerOIDCSubject *string
	OwnerAPIKeyID    *uuid.UUID
	CreatedAt        time.Time
	UpdatedAt        time.Time
}

// DispatchUnit is the in-memory view of a dispatch_units row. Each unit
// addresses exactly one spec file; the orchestrator does not bundle specs.
type DispatchUnit struct {
	ID             uuid.UUID
	RunID          uuid.UUID
	DispatchSeq    int
	SpecPath       string
	State          string
	CurrentLeaseID *uuid.UUID
	LeaseCount     int
	FailCount      int
	OutcomeSetAt   *time.Time
}

// Lease is the in-memory view of a leases row.
type Lease struct {
	ID              uuid.UUID
	RunID           uuid.UUID
	Worker          WorkerIdentity
	UnitIDs         []uuid.UUID
	IssuedAt        time.Time
	Deadline        time.Time
	ReleasedAt      *time.Time
	ReleaseReason   *string
	AuthOIDCSubject *string
	AuthAPIKeyID    *uuid.UUID
}

// Attempt is the in-memory view of an attempts row.
type Attempt struct {
	ID               uuid.UUID
	LeaseID          uuid.UUID
	DispatchUnitID   uuid.UUID
	RunID            uuid.UUID
	SpecPath         string
	Status           *string
	ActualDurationMs *int64
	ErrorMessage     *string
	ErrorStack       *string
	TestCases        json.RawMessage
	ReportedAt       *time.Time
	LateReport       bool
	Expired          bool
	CreatedAt        time.Time
}

// AttemptView is an Attempt denormalized with the worker name from the
// joined leases row. Exposed in the run-status payload so the dashboard
// can show "spec X was run by worker Y" without a per-row lookup.
type AttemptView struct {
	Attempt
	GHJobName string
	GHJobID   string
}

// CurrentLeaseView captures the minimal lease info shown on a unit while it
// is in the leased state. Populated only when DispatchUnit.CurrentLeaseID is
// non-nil; nil otherwise.
type CurrentLeaseView struct {
	ID        uuid.UUID
	GHJobName string
	GHJobID   string
	IssuedAt  time.Time
	Deadline  time.Time
}

// UnitView bundles a DispatchUnit with the data the dashboard needs to
// render its current state: the active lease (when leased) and the full
// attempt history. Attempts are ordered oldest-first.
type UnitView struct {
	DispatchUnit
	CurrentLease *CurrentLeaseView
	Attempts     []AttemptView
}

// RunWithUnits is the full read-model returned by Store.GetRunWithUnits and
// serialized into the /orchestration/status and /orchestration/begin
// response bodies. Units are ordered by dispatch_seq.
type RunWithUnits struct {
	Run   *Run
	Units []UnitView
}

// Validate enforces required fields and the framework pin on a composite
// identity. It does NOT mutate (Framework defaulting is the caller's job —
// typically the HTTP handler).
func (ci CompositeIdentity) Validate() error {
	if ci.Repository == "" {
		return errors.New("composite identity: repository is required")
	}
	if ci.CommitSHA == "" {
		return errors.New("composite identity: commit_sha is required")
	}
	if ci.GHRunID == "" {
		return errors.New("composite identity: gh_run_id is required")
	}
	if ci.Name == "" {
		return errors.New("composite identity: name is required")
	}
	if ci.GHRunAttempt == "" {
		return errors.New("composite identity: gh_run_attempt is required")
	}
	if !IsSupportedFramework(ci.Framework) {
		return fmt.Errorf("composite identity: framework %q is not supported (must be one of: %s)", ci.Framework, strings.Join(SupportedFrameworksList(), ", "))
	}
	return nil
}

// HashUnits returns a SHA-256 digest of the canonical-form serialization of
// the dispatch-units list. Canonical form: the spec_path strings are kept
// in submission order (no sort) and JSON-marshaled, then hashed. Used by
// the begin-run idempotency check — a retry whose hash matches the stored
// value is treated as a no-op replay; a retry with a different hash on
// the same composite identity is a conflict.
//
// Note: caller submission order is part of the canonical form because
// dispatch_seq drives FIFO checkout — re-ordering changes run behavior
// and must therefore be a different run.
func (ci CompositeIdentity) HashUnits(specPaths []string) []byte {
	buf, _ := json.Marshal(specPaths)
	sum := sha256.Sum256(buf)
	return sum[:]
}
